File: Arrays/Inplace/SortByParity.cpp
class Solution {
public:
  vector<int> sortArrayByParity(vector<int>& nums) {
    int len = nums.size(), start = 0, end = len - 1;
    while(start < end){
      if(!(nums[start] % 2)) // if start one is even, skip to next one
        start++;
      else {
        // if odd was found at the start, check if even is found at the end
        if(!(nums[end] % 2)){
          // even found at the end, replace
          int temp = nums[start];
          nums[start] = nums[end];
          nums[end] = temp;
        }
        // odd is found in the beginning and odd is found at the end, let the odd be at the end
        end--;
      }
    }
    return nums;
  }
};

File: Arrays/Inplace/SortByParity.js
// https://leetcode.com/explore/learn/card/fun-with-arrays/511/in-place-operations/3260/
/**
 * @param {number[]} A
 * @return {number[]}
 */
const sortArrayByParity = function (A) {
  const LEN = A.length;
  let start = 0, end = LEN - 1;
  while (start < end) {
    if (A[start] % 2 == 0) // check if start one is even
      start++;
    else {
      if (A[end] % 2 == 0) { // even
        let temp = A[start];
        A[start] = A[end];
        A[end] = temp;
      }
      end--;
    }
  }
  return A;
};

const arr = [3, 1, 2, 4];
const res = sortArrayByParity(arr);
console.log("Response:", res);